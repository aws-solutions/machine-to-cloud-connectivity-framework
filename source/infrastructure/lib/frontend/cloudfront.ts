// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0

import { CustomResource, CfnCondition, CfnCustomResource, Aws } from 'aws-cdk-lib';
import { CloudFrontToS3 } from '@aws-solutions-constructs/aws-cloudfront-s3';
import { SecurityPolicyProtocol } from 'aws-cdk-lib/aws-cloudfront';
import { Bucket, IBucket } from 'aws-cdk-lib/aws-s3';
import { NagSuppressions } from 'cdk-nag';
import { Construct } from 'constructs';

export interface CloudFrontConstructProps {
  readonly s3LoggingBucket: IBucket;
  readonly customResourcesFunctionArn: string;
  readonly shouldTeardownData: CfnCondition;
}

/**
 * Creates a CloudFront distribution.
 */
export class CloudFrontConstruct extends Construct {
  public cloudFrontDomainName: string;
  public uiBucket: Bucket;

  constructor(scope: Construct, id: string, props: CloudFrontConstructProps) {
    super(scope, id);

    const cloudFrontToS3 = new CloudFrontToS3(this, 'CloudFrontToS3', {
      bucketProps: {
        serverAccessLogsBucket: props.s3LoggingBucket,
        serverAccessLogsPrefix: 'ui-s3/',
        bucketName: `${Aws.STACK_NAME}-${Aws.ACCOUNT_ID}-ui`
      },
      cloudFrontDistributionProps: {
        comment: 'Machine to Cloud Connectivity Framework Distribution',
        enableLogging: true,
        errorResponses: [
          { httpStatus: 403, responseHttpStatus: 200, responsePagePath: '/index.html' },
          { httpStatus: 404, responseHttpStatus: 200, responsePagePath: '/index.html' }
        ],
        logBucket: props.s3LoggingBucket,
        minimumProtocolVersion: SecurityPolicyProtocol.TLS_V1_2_2019,
        logFilePrefix: 'ui-cf/'
      },
      insertHttpSecurityHeaders: false
    });
    this.cloudFrontDomainName = cloudFrontToS3.cloudFrontWebDistribution.domainName;
    this.uiBucket = <Bucket>cloudFrontToS3.s3Bucket;

    const teardownCloudfrontBucket = new CustomResource(this, 'TeardownCloudfrontBucket', {
      serviceToken: props.customResourcesFunctionArn,
      properties: {
        Resource: 'DeleteS3Bucket',
        BucketName: this.uiBucket.bucketName
      }
    });
    const cfnTeardownCloudfrontBucket = <CfnCustomResource>teardownCloudfrontBucket.node.defaultChild;
    cfnTeardownCloudfrontBucket.cfnOptions.condition = props.shouldTeardownData;

    NagSuppressions.addResourceSuppressions(
      cloudFrontToS3,
      [
        { id: 'AwsSolutions-CFR1', reason: 'The solution does not control geo restriction.' },
        { id: 'AwsSolutions-CFR2', reason: 'No need to enable WAF.' },
        {
          id: 'AwsSolutions-CFR4',
          reason: 'No contorl on the solution side as it is using the CloudFront default certificate.'
        }
      ],
      true
    );
  }
}
