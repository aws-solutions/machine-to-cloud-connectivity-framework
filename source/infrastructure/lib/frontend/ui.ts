// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0

import { Aws, Duration, RemovalPolicy, Stack } from 'aws-cdk-lib';
import {
  CfnIdentityPool,
  CfnIdentityPoolRoleAttachment,
  CfnUserPool,
  CfnUserPoolUser,
  UserPool,
  UserPoolClient
} from 'aws-cdk-lib/aws-cognito';
import { Effect, FederatedPrincipal, PolicyDocument, PolicyStatement, Role } from 'aws-cdk-lib/aws-iam';
import { CfnBucket, IBucket } from 'aws-cdk-lib/aws-s3';
import { NagSuppressions } from 'cdk-nag';
import { Construct } from 'constructs';

export interface UiConstructProps {
  readonly apiId: string;
  readonly resourceBucket: IBucket;
  readonly userEmail: string;
  readonly cloudFrontDomainName: string;
}

/**
 * Creates a CloudFront distribution, an UI hosting S3 bucket, and Cognito resources.
 */
export class UiConstruct extends Construct {
  public identityPoolId: string;
  public userPoolId: string;
  public webClientId: string;

  constructor(scope: Construct, id: string, props: UiConstructProps) {
    super(scope, id);

    const userPool = new UserPool(this, 'UserPool', {
      passwordPolicy: {
        minLength: 12,
        requireDigits: true,
        requireLowercase: true,
        requireSymbols: true,
        requireUppercase: true
      },
      removalPolicy: RemovalPolicy.DESTROY,
      selfSignUpEnabled: false,
      signInAliases: {
        email: true
      },
      userPoolName: `${Aws.STACK_NAME}-user-pool`,
      userInvitation: {
        emailSubject: '[Machine to Cloud Connectivity Framework] Login information',
        emailBody: `
          <p>
            You are invited to join Machine to Cloud Connectivity Framework.<br />
            https://${props.cloudFrontDomainName}
          </p>
          <p>
            Please sign in to Machine to Cloud Connectivity Framework using the temporary credentials below:<br />
            Username: <strong>{username}</strong><br />Password: <strong>{####}</strong>
          </p>
        `
      }
    });
    (<CfnUserPool>userPool.node.defaultChild).userPoolAddOns = { advancedSecurityMode: 'ENFORCED' };
    this.userPoolId = userPool.userPoolId;

    const userPoolClient = new UserPoolClient(this, 'UserPoolClient', {
      generateSecret: false,
      preventUserExistenceErrors: true,
      refreshTokenValidity: Duration.days(1),
      userPool,
      userPoolClientName: `${Aws.STACK_NAME}-userpool-client`
    });
    this.webClientId = userPoolClient.userPoolClientId;

    new CfnUserPoolUser(this, 'User', {
      userPoolId: userPool.userPoolId,
      desiredDeliveryMediums: ['EMAIL'],
      forceAliasCreation: true,
      userAttributes: [
        { name: 'email', value: props.userEmail },
        { name: 'email_verified', value: 'true' }
      ],
      username: props.userEmail
    });

    const identityPool = new CfnIdentityPool(this, 'IdentityPool', {
      allowUnauthenticatedIdentities: false,
      cognitoIdentityProviders: [
        {
          clientId: userPoolClient.userPoolClientId,
          providerName: userPool.userPoolProviderName,
          serverSideTokenCheck: false
        }
      ]
    });
    this.identityPoolId = identityPool.ref;

    const authenticatedRole = new Role(this, 'IdentityPoolAuthenticatedRole', {
      assumedBy: new FederatedPrincipal(
        'cognito-identity.amazonaws.com',
        {
          StringEquals: { 'cognito-identity.amazonaws.com:aud': identityPool.ref },
          'ForAnyValue:StringLike': { 'cognito-identity.amazonaws.com:amr': 'authenticated' }
        },
        'sts:AssumeRoleWithWebIdentity'
      ),
      description: `${Aws.STACK_NAME} Identity Pool authenticated role`,
      inlinePolicies: {
        ExecuteApiPolicy: new PolicyDocument({
          statements: [
            new PolicyStatement({
              effect: Effect.ALLOW,
              actions: ['execute-api:Invoke'],
              resources: [
                Stack.of(this).formatArn({
                  service: 'execute-api',
                  resource: props.apiId,
                  resourceName: 'prod/*'
                })
              ]
            }),
            new PolicyStatement({
              effect: Effect.ALLOW,
              actions: ['s3:GetObject'],
              resources: [props.resourceBucket.arnForObjects('public/*')]
            })
          ]
        })
      }
    });

    new CfnIdentityPoolRoleAttachment(this, 'IdentityPoolRoleAttachment', {
      identityPoolId: identityPool.ref,
      roles: { authenticated: authenticatedRole.roleArn }
    });

    (props.resourceBucket.node.defaultChild as CfnBucket).corsConfiguration = {
      corsRules: [
        {
          allowedMethods: ['GET'],
          allowedOrigins: [`https://${props.cloudFrontDomainName}`],
          allowedHeaders: ['*'],
          exposedHeaders: ['ETag']
        }
      ]
    };

    NagSuppressions.addResourceSuppressions(userPool, [
      { id: 'AwsSolutions-COG2', reason: 'No need to enable MFA as that is option for users.' }
    ]);
    NagSuppressions.addResourceSuppressions(authenticatedRole, [
      { id: 'AwsSolutions-IAM5', reason: 'It does not allow wildcard permissions.' }
    ]);
  }
}
