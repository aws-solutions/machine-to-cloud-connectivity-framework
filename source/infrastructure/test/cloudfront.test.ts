// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0

import '@aws-cdk/assert/jest';
import { SynthUtils } from '@aws-cdk/assert';
import { CfnCondition, Stack } from 'aws-cdk-lib';
import { Bucket, CfnBucket } from 'aws-cdk-lib/aws-s3';
import { CloudFrontConstruct } from '../lib/frontend/cloudfront';

test('M2C2 Cloudfront test', () => {
  const stack = new Stack();
  const s3LoggingBucket = new Bucket(stack, 'TestLoggingBucket');
  (<CfnBucket>s3LoggingBucket.node.defaultChild).overrideLogicalId('TestLoggingBucket');
  const resourceBucket = new Bucket(stack, 'TestGreengrassResourceBucket');
  (<CfnBucket>resourceBucket.node.defaultChild).overrideLogicalId('TestGreengrassResourceBucket');

  const cf = new CloudFrontConstruct(stack, 'TestCF', {
    s3LoggingBucket: s3LoggingBucket,
    customResourcesFunctionArn: 'test-arn',
    shouldTeardownData: new CfnCondition(stack, 'TestCondition')
  });

  expect(SynthUtils.toCloudFormation(stack)).toMatchSnapshot();
  expect(cf.uiBucket).toBeDefined();
  expect(cf.cloudFrontDomainName).toBeDefined();
  expect(stack).toHaveResourceLike('AWS::S3::Bucket', {
    LoggingConfiguration: {
      DestinationBucketName: {
        Ref: 'TestLoggingBucket'
      },
      LogFilePrefix: 'ui-s3/'
    }
  });
  expect(stack).toHaveResourceLike('AWS::CloudFront::Distribution', {
    DistributionConfig: {
      Logging: {
        Bucket: {
          'Fn::GetAtt': ['TestLoggingBucket', 'RegionalDomainName']
        },
        Prefix: 'ui-cf/'
      }
    }
  });
});
