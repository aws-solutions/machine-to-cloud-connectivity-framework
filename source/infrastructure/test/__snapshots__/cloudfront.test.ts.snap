// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`M2C2 Cloudfront test 1`] = `
Object {
  "Resources": Object {
    "TestCFCloudFrontToS3CloudFrontDistribution37BBD3EE": Object {
      "Metadata": Object {
        "cdk_nag": Object {
          "rules_to_suppress": Array [
            Object {
              "id": "AwsSolutions-CFR1",
              "reason": "The solution does not control geo restriction.",
            },
            Object {
              "id": "AwsSolutions-CFR2",
              "reason": "No need to enable WAF.",
            },
            Object {
              "id": "AwsSolutions-CFR4",
              "reason": "No contorl on the solution side as it is using the CloudFront default certificate.",
            },
          ],
        },
        "cfn_nag": Object {
          "rules_to_suppress": Array [
            Object {
              "id": "W70",
              "reason": "Since the distribution uses the CloudFront domain name, CloudFront automatically sets the security policy to TLSv1 regardless of the value of MinimumProtocolVersion",
            },
          ],
        },
      },
      "Properties": Object {
        "DistributionConfig": Object {
          "Comment": "Machine to Cloud Connectivity Framework Distribution",
          "CustomErrorResponses": Array [
            Object {
              "ErrorCode": 403,
              "ResponseCode": 200,
              "ResponsePagePath": "/index.html",
            },
            Object {
              "ErrorCode": 404,
              "ResponseCode": 200,
              "ResponsePagePath": "/index.html",
            },
          ],
          "DefaultCacheBehavior": Object {
            "CachePolicyId": "658327ea-f89d-4fab-a63d-7e88639e58f6",
            "Compress": true,
            "TargetOriginId": "TestCFCloudFrontToS3CloudFrontDistributionOrigin19002C6E4",
            "ViewerProtocolPolicy": "redirect-to-https",
          },
          "DefaultRootObject": "index.html",
          "Enabled": true,
          "HttpVersion": "http2",
          "IPV6Enabled": true,
          "Logging": Object {
            "Bucket": Object {
              "Fn::GetAtt": Array [
                "TestLoggingBucket",
                "RegionalDomainName",
              ],
            },
            "Prefix": "ui-cf/",
          },
          "Origins": Array [
            Object {
              "DomainName": Object {
                "Fn::GetAtt": Array [
                  "TestCFCloudFrontToS3S3Bucket2764548A",
                  "RegionalDomainName",
                ],
              },
              "Id": "TestCFCloudFrontToS3CloudFrontDistributionOrigin19002C6E4",
              "S3OriginConfig": Object {
                "OriginAccessIdentity": Object {
                  "Fn::Join": Array [
                    "",
                    Array [
                      "origin-access-identity/cloudfront/",
                      Object {
                        "Ref": "TestCFCloudFrontToS3CloudFrontDistributionOrigin1S3Origin9082F7A6",
                      },
                    ],
                  ],
                },
              },
            },
          ],
        },
      },
      "Type": "AWS::CloudFront::Distribution",
    },
    "TestCFCloudFrontToS3CloudFrontDistributionOrigin1S3Origin9082F7A6": Object {
      "Metadata": Object {
        "cdk_nag": Object {
          "rules_to_suppress": Array [
            Object {
              "id": "AwsSolutions-CFR1",
              "reason": "The solution does not control geo restriction.",
            },
            Object {
              "id": "AwsSolutions-CFR2",
              "reason": "No need to enable WAF.",
            },
            Object {
              "id": "AwsSolutions-CFR4",
              "reason": "No contorl on the solution side as it is using the CloudFront default certificate.",
            },
          ],
        },
      },
      "Properties": Object {
        "CloudFrontOriginAccessIdentityConfig": Object {
          "Comment": "Identity for TestCFCloudFrontToS3CloudFrontDistributionOrigin19002C6E4",
        },
      },
      "Type": "AWS::CloudFront::CloudFrontOriginAccessIdentity",
    },
    "TestCFCloudFrontToS3S3Bucket2764548A": Object {
      "DeletionPolicy": "Retain",
      "Metadata": Object {
        "cdk_nag": Object {
          "rules_to_suppress": Array [
            Object {
              "id": "AwsSolutions-CFR1",
              "reason": "The solution does not control geo restriction.",
            },
            Object {
              "id": "AwsSolutions-CFR2",
              "reason": "No need to enable WAF.",
            },
            Object {
              "id": "AwsSolutions-CFR4",
              "reason": "No contorl on the solution side as it is using the CloudFront default certificate.",
            },
          ],
        },
      },
      "Properties": Object {
        "BucketEncryption": Object {
          "ServerSideEncryptionConfiguration": Array [
            Object {
              "ServerSideEncryptionByDefault": Object {
                "SSEAlgorithm": "AES256",
              },
            },
          ],
        },
        "BucketName": Object {
          "Fn::Join": Array [
            "",
            Array [
              Object {
                "Ref": "AWS::StackName",
              },
              "-",
              Object {
                "Ref": "AWS::AccountId",
              },
              "-ui",
            ],
          ],
        },
        "LifecycleConfiguration": Object {
          "Rules": Array [
            Object {
              "NoncurrentVersionTransitions": Array [
                Object {
                  "StorageClass": "GLACIER",
                  "TransitionInDays": 90,
                },
              ],
              "Status": "Enabled",
            },
          ],
        },
        "LoggingConfiguration": Object {
          "DestinationBucketName": Object {
            "Ref": "TestLoggingBucket",
          },
          "LogFilePrefix": "ui-s3/",
        },
        "PublicAccessBlockConfiguration": Object {
          "BlockPublicAcls": true,
          "BlockPublicPolicy": true,
          "IgnorePublicAcls": true,
          "RestrictPublicBuckets": true,
        },
        "VersioningConfiguration": Object {
          "Status": "Enabled",
        },
      },
      "Type": "AWS::S3::Bucket",
      "UpdateReplacePolicy": "Retain",
    },
    "TestCFCloudFrontToS3S3BucketPolicy111FF803": Object {
      "Metadata": Object {
        "cdk_nag": Object {
          "rules_to_suppress": Array [
            Object {
              "id": "AwsSolutions-CFR1",
              "reason": "The solution does not control geo restriction.",
            },
            Object {
              "id": "AwsSolutions-CFR2",
              "reason": "No need to enable WAF.",
            },
            Object {
              "id": "AwsSolutions-CFR4",
              "reason": "No contorl on the solution side as it is using the CloudFront default certificate.",
            },
          ],
        },
        "cfn_nag": Object {
          "rules_to_suppress": Array [
            Object {
              "id": "F16",
              "reason": "Public website bucket policy requires a wildcard principal",
            },
          ],
        },
      },
      "Properties": Object {
        "Bucket": Object {
          "Ref": "TestCFCloudFrontToS3S3Bucket2764548A",
        },
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "s3:*",
              "Condition": Object {
                "Bool": Object {
                  "aws:SecureTransport": "false",
                },
              },
              "Effect": "Deny",
              "Principal": Object {
                "AWS": "*",
              },
              "Resource": Array [
                Object {
                  "Fn::GetAtt": Array [
                    "TestCFCloudFrontToS3S3Bucket2764548A",
                    "Arn",
                  ],
                },
                Object {
                  "Fn::Join": Array [
                    "",
                    Array [
                      Object {
                        "Fn::GetAtt": Array [
                          "TestCFCloudFrontToS3S3Bucket2764548A",
                          "Arn",
                        ],
                      },
                      "/*",
                    ],
                  ],
                },
              ],
            },
            Object {
              "Action": "s3:GetObject",
              "Effect": "Allow",
              "Principal": Object {
                "CanonicalUser": Object {
                  "Fn::GetAtt": Array [
                    "TestCFCloudFrontToS3CloudFrontDistributionOrigin1S3Origin9082F7A6",
                    "S3CanonicalUserId",
                  ],
                },
              },
              "Resource": Object {
                "Fn::Join": Array [
                  "",
                  Array [
                    Object {
                      "Fn::GetAtt": Array [
                        "TestCFCloudFrontToS3S3Bucket2764548A",
                        "Arn",
                      ],
                    },
                    "/*",
                  ],
                ],
              },
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::S3::BucketPolicy",
    },
    "TestCFTeardownCloudfrontBucket814A2F37": Object {
      "Condition": "TestCondition",
      "DeletionPolicy": "Delete",
      "Properties": Object {
        "BucketName": Object {
          "Ref": "TestCFCloudFrontToS3S3Bucket2764548A",
        },
        "Resource": "DeleteS3Bucket",
        "ServiceToken": "test-arn",
      },
      "Type": "AWS::CloudFormation::CustomResource",
      "UpdateReplacePolicy": "Delete",
    },
    "TestGreengrassResourceBucket": Object {
      "DeletionPolicy": "Retain",
      "Type": "AWS::S3::Bucket",
      "UpdateReplacePolicy": "Retain",
    },
    "TestLoggingBucket": Object {
      "DeletionPolicy": "Retain",
      "Properties": Object {
        "AccessControl": "LogDeliveryWrite",
      },
      "Type": "AWS::S3::Bucket",
      "UpdateReplacePolicy": "Retain",
    },
  },
}
`;
